src/main/asciidoc/repository-projections.adoc

```diff
@@ -1,163 +1,276 @@
+Methods backed by SpEL expression evaluation can also use method parameters which can then be referred to from the expression.
+The method parameters are available via an `Object` array named `args`.
 
-  @Value("#{target.address.country}")
-  String getCountry();
+.Sample Person object
+====
+[source, java]
+----
+interface NamesOnly {
+
+  @Value("#{args[0] + ' ' + target.firstname + '!'}")
+  String getSalutation(String prefix);
 }
 ----
+====
 
-In fact, `@Value` gives full access to the target object and its nested properties. SpEL expressions are extremly powerful as the definition is always applied to the projection method. Let's take SpEL expressions in projections to the next level.
+Again, for more complex expressions rather use a Spring bean and let the expression just invoke a method as described in <<projections.interfaces.open.bean-reference, above>>.
 
+[[projections.dtos]]
+== Class-based projections (DTOs)
 
-Imagine you had the following domain model definition:
+Another way of defining
 
-[source,java]
+.A projecting DTO
+====
+[source, java]
 ----
-@Entity
-public class User {
+class NamesOnly {
 
-  @Id @GeneratedValue
-  private Long id;
-  private String name;
+  private final String firstname, lastname;
 
-  private String password;
-  …
+  NamesOnly(String firstname, String lastname) {
+    this.firstname = firstname;
+    this.lastname = lastname;
+  }
+
+  String getFirstname() {
+    return this.firstname;
+  }
+
+  String getLastname() {
+    return this.lastname;
+  }
+
+  // equals(…) and hashCode() implementations
+}
+----
+====
+
+[TIP]
+.Avoiding boilerplate code for projection DTOs
+====
+The code that needs to be written for a DTO can be dramatically simplified using https://projectlombok.org[Project Lombok], which provides an `@Value` annotation (not to mix up with Spring's `@Value` annotation shown in the interface examples above).
+The sample DTO above would become this:
+
+[source, java]
+----
+@Value
+class NamesOnly {
+	String firstname, lastname;
 }
 ----
+Fields are private final by default, the class exposes a constructor taking all fields and automatically gets `equals(…)` and `hashCode()` methods implemented.
 
-IMPORTANT: This example may seem a bit contrived, but it is possible with a richer domain model and many projections, to accidentally leak such details. Since Spring Data cannot discern the sensitivity of such data, it is up to the developers to avoid such situations. Storing a password as plain-text is discouraged. You really should not do this. For this example, you could also replace `password` with anything else that is secret.
+====
+
+[[projection.dynamic]]
+== Dynamic projections
 
-In some cases, you might keep the `password` as secret as possible and not expose it more than it should be. The solution is to create a projection using `@Value` together with a SpEL expression.
+So far we have seen the projection type to be used declared as return type or element type of a collection.
```

Comment:
A little simpler:
So far we have declared the projection type as the return type or element type of a collection.
Or:
So far we have used the projection type as the return type or element type of a collection.